### DISCUSSION TITLE
Riok.Mapperly: 3.1.0 : CS8795 Partial method 'xxx' must have an implementation part because it has accessibility modifiers.

### DISCUSSION THREAD

**Discussion by senal** [riok/mapperly#663] [Q&A] [answered]
Dears,

I'm trying out mapperly in one of my projects.
I have an implementation like below. But I'm getting this IDE design time error `CS8795 Partial method 'xxx' must have an implementation part because it has accessibility modifiers.` on  line ` public static partial Vehicle MapExternalVehicleToVehicle(...)` 

![image](https://github.com/riok/mapperly/assets/2841066/5c53ceaa-6280-40c1-8e5f-4dbb6dd378a2)

    public static partial class VehicleMapper
    {
         // map properties with different names
         [MapProperty(nameof(ExternalVehicle.Make_Year), nameof(Vehicle.MakeYear))]
        
        // map nested props (Engine)
        [MapProperty(nameof(ExternalVehicle.Engine.Engine_Liters), nameof(Vehicle.Engine.Liters))]
        
        // map nested props (Registration)
        [MapProperty(nameof(ExternalVehicle.Registration.Registration_Status), nameof(Vehicle.Registration.Status))]
        
       public static partial Vehicle MapExternalVehicleToVehicle(ExternalVehicle externalVehicle);
   }

Any feedback/comment is appreciated. 

Thanks

**Comment by TimothyMakkison** [comment]
Hey, have you tried restarting your ide? Source generators sometimes need a restart before working

Edit: I reread your code and it looks like your `MapProperty` are invalid.

```C#
// map nested props (Engine)
[MapProperty(nameof(ExternalVehicle.Engine.Engine_Liters), nameof(Vehicle.Engine.Liters))]
    
// map nested props (Registration)
[MapProperty(nameof(ExternalVehicle.Registration.Registration_Status), nameof(Vehicle.Registration.Status))]    
```
```C#
// map nested props (Engine)
[MapProperty("Engine.Engine_Liters", "Engine.Liters")]
        
// map nested props (Registration)
[MapProperty("Registration.Registration_Status", "Registration.Status")]   
```
Your error is due to a common misconception around `nameof`. `nameof` will only return the final member and will not print the preceeding ones. `nameof(ExternalVehicle.Engine.Engine_Liters)` will return `"Engine_Liters"` instead of `"Engine.Engine_Liters"`

You should have some other warnings about Mapperly being unable to find various properties.

**Reply by senal** [reply]
@TimothyMakkison ,
Thank you very much for your prompt feedback,
I did all solutions listed in #519
Thank you very much for pointing me in the right direction of the usage of `nameof(...)`.
I did change the mapping as you suggested.
And tried again restarting the IDE a couple of times. I'm using VS. I even upgrade the IDE to the latest version 17.7.1.
Even I tried downgrading the Mapperly to 3.0.0
Nothing worked for me.

Changed implementation as per your feedback.

    public static partial class VehicleMapper
    {
        // map properties with different names
        [MapProperty("ExternalVehicle.Make_Year", "Vehicle.MakeYear")]

        // map nested props (Engine)
        [MapProperty("ExternalVehicle.Engine.Engine_Liters", "Vehicle.Engine.Liters")]
        [MapProperty("ExternalVehicle.Engine.Engine_Cylinders", "Vehicle.Engine.Cylinders")]
        [MapProperty("ExternalVehicle.Engine.Engine_Number", "Vehicle.Engine.Number")]

        // map nested props (Registration)
        [MapProperty("ExternalVehicle.Registration.Registration_Status", "Vehicle.Registration.Status")]
        [MapProperty("ExternalVehicle.Registration.Registration_State", "Vehicle.Registration.State")]
        public static partial Vehicle MapExternalVehicleToVehicle(ExternalVehicle externalVehicle);
    }

I did write another sample mapping using the `Mapper`  attribute at the class level as follows. but this one is a direct mapping and no nested props involved as in my previous example.

    [Mapper]
    public static partial class CarMapper
    {
        public static partial CarDto MapCarToCarDto(Car car);
    }

And that generates the targeted mapping as follows.

    public static partial class CarMapper
    {
        public static partial global::TryMapperly.Infra.CarDto MapCarToCarDto(global::TryMapperly.Infra.Car car)
        {
            var target = new global::TryMapperly.Infra.CarDto()
            {
                Modal = car.Modal,
                Make = car.Make,
                Manufacture_Year = car.Manufacture_Year,
                Build_Year = car.Build_Year
            };
            return target;
        }
    }

Please let me know if you require me to change or try out another option
Appreciate your feedback.

Thanks

**Comment by CommonGuy** [comment]
Did you add the `[Mapper]` attribute to your VehicleMapper class?

**Comment by TimothyMakkison** [comment, marked as answer]
> Even I tried downgrading the Mapperly to 3.0.0

Use `3.1.0` instead, `3.0.0` is broken for  VS users 😞 

Try this:
```C#
[Mapper]
public static partial class VehicleMapper
{
    // map properties with different names
    [MapProperty("Make_Year", "MakeYear")]

    // map nested props (Engine)
    [MapProperty("Engine.Engine_Liters", "Engine.Liters")]
    [MapProperty("Engine.Engine_Cylinders", "Engine.Cylinders")]
    [MapProperty("Engine.Engine_Number", "Engine.Number")]

    // map nested props (Registration)
    [MapProperty("Registration.Registration_Status", "Registration.Status")]
    [MapProperty("Registration.Registration_State", "Registration.State")]
    public static partial Vehicle MapExternalVehicleToVehicle(ExternalVehicle externalVehicle);
}
```

i removed the unneeded types from the start of `MapProperty`
`MapProperty("ExternalVehicle.Registration.Registration_Status", "Vehicle.Registration.Status")` will try to find a member `ExternalVehicle` with the sub members `Registration.Registration_Status` whereas you want to map the member `Registration` with the sub member `Registration_Status`.

**Comment by senal** [comment]
@TimothyMakkison ,
Your solution fixed my issue.
But I had to initialize the `Engine` and `Registration` within the `Vehicle` constructor. (I was under the impression that Mapperly initiated the nested props which have complex objects).

Below is how the mapping and the `vehicle` classes are now look like

Also, I have a question about `set` vs `init`.  
As you can see some of the props in `Vehicle` class mark with `init`  but in the nested props (`Engine` and `Registration`) I had to leave them with `get and set` ( including props belonging to `Engine` and `Registration` because Mapperly failed to map the values to the props.
Do you have any idea why the props with `init` in nested props didn't get the mapped value? Is it a bug? 
Appreciate your feedback. 

    [Mapper]
    public static partial class VehicleMapper
    {
        // map properties with different names
        [MapProperty("Make_Year", "MakeYear")]

        // map nested props (Engine)
        [MapProperty("Engine.Engine_Liters", "Engine.Liters")]
        [MapProperty("Engine.Engine_Cylinders", "Engine.Cylinders")]
        [MapProperty("Engine.Engine_Number", "Engine.Number")]

        // map nested props (Registration)
        [MapProperty("Registration.Registration_Plate", "Registration.Plate")]
        [MapProperty("Registration.Registration_Status", "Registration.Status")]
        [MapProperty("Registration.Registration_State", "Registration.State")]
        public static partial Vehicle MapExternalVehicleToVehicle(ExternalVehicle externalVehicle);
    }

Vehicle class

    public class Vehicle
    {
	public Vehicle( )
	{
		Engine = new Engine();
		Registration = new Registration();
	}
	public string Make { get; init; } = default!;
	public string Model { get; init; } = default!;
	public int MakeYear { get; init; } = default!;
	public Engine Engine { get; set; }
	public Registration Registration { get; set; }

    }
    public class Engine
    {
           public int Cylinders { get; set; } = default!;
	    public int Liters { get; set; } = default!;
	    public int Number { get; set; } = default!;
     }
    public class Registration
    {
        public string Plate { get; set; } = default!;
        public string Status { get; set; } = default!;
        public string State { get; set; } = default!;
    }

Thanks
